src/Dialect/ONNX/ONNXEinsumOpHelper.cpp

```diff
@@ -0,0 +1,319 @@
+  // iterate through sorted map in order, i.e. alphabetically and
+  // all upper case letters before lower case
+  for (const auto &entry : counts) { // entry == pair (x, count)
+    if (entry.second == 1)           // one occurrence
+      equationOutput.push_back(entry.first);
+  }
+  return equationOutput;
+}
+
+// Argument must be an input or output parameter consisting of letters,
+// spaces, and possibly ellipsis.
+size_t countLetters(StringRef parameter) {
+  return parameter.size() - parameter.count(' ') - parameter.count('.');
+}
+
+bool hasEllipsis(StringRef parameter) {
+  return parameter.find('.') != StringRef::npos;
+}
+
+void appendLetterSubscripts(StringRef letters, Subscripts &subscripts) {
+  size_t p = letters.find_first_not_of(' ');
+  while (p != StringRef::npos) {
+    subscripts.push_back(letters[p]);
+    p = letters.find_first_not_of(' ', p + 1);
+  }
+}
+
+// It's convenient to limit the max ellipsis rank to 10 so we can represent
+// an ellipsis internally as a sequence of digit subscripts which is easy
+// to read when debugging.
+// (Easy to change this if there's ever a use case for larger ellipsis rank.)
+constexpr int64_t kMaxEllipsisRank = 10;
+
+// We represent an ellipsis as a sequence of digit subscripts, one per axis.
+char ellipsisSubscript(int64_t axis) {
+  assert(0 <= axis && axis < 10);
+  return '0' + axis;
+}
+
+Subscripts extractSubscripts(StringRef parameterEquation, int64_t rank) {
+  Subscripts subscripts;
+  subscripts.reserve(rank);
+  StringRef prefix, suffix;
+  std::tie(prefix, suffix) = parameterEquation.split('.');
+  int64_t prefixRank = prefix.size() - prefix.count(' ');
+  appendLetterSubscripts(prefix, subscripts);
+  if (prefix.size() == parameterEquation.size()) {
+    assert(prefixRank == rank);
+  } else {
+    // Skip past remainder of ellipsis "..." and any spaces
+    suffix = suffix.ltrim(" .");
+    int64_t suffixRank = suffix.size() - suffix.count(' ');
+    assert(rank >= prefixRank + suffixRank);
+    int64_t ellipsisRank = rank - prefixRank - suffixRank;
+    for (int64_t axis = 0; axis < ellipsisRank; ++axis) {
+      subscripts.push_back(ellipsisSubscript(axis));
+    }
+    appendLetterSubscripts(suffix, subscripts);
+  }
+  assert((int64_t)subscripts.size() == rank);
+  return subscripts;
+}
+
+} // namespace
+
+// Preconditions:
+// . equation must be well-formed as validated by verifyEquation()
+// . number of equation input parameters must match number of actual inputs
+// . all input types must be shaped types with rank
+FailureOr<Signature> inferSignature(
+    ONNXEinsumOpAdaptor operandAdaptor, ErrorFn emitErrorFn) {
+  auto equation = operandAdaptor.equation();
+  auto inputs = operandAdaptor.Inputs();
+  assert(succeeded(verifyEquation(equation, inputs.size(),
+      emitErrorFn))); // precondition, TODO: remove this excessive check
+  assert(llvm::all_of(inputs, [](Value i) {
+    return i.getType().cast<ShapedType>().hasRank();
```

Comment:
I used isRankedShapedType from TypeUtilities, thanks for the suggestion

FWIW, I gather it isn't really needed to check if types are ShapedType here because the EInsum spec says that all types are tensors and that's checked in ONNXEinsumOp::verifyInvariants()